module/VuFind/src/VuFind/AjaxHandler/GetSearchResults.php

```diff
@@ -0,0 +1,455 @@
+            if (empty($driverAndKeys['keys'])) {
+                return null;
+            }
+
+            $searchParams['lookfor'] = $this->versionsHelper
+                ->getSearchStringFromWorkKeys((array)$driverAndKeys['keys']);
+            $searchParams['type'] = $this->versionsHelper->getWorkKeysSearchType();
+        }
+
+        $results = $this->resultsManager->get($backend);
+        $paramsObj = $results->getParams();
+        $paramsObj->getOptions()->spellcheckEnabled(false);
+        $paramsObj->initFromRequest(new Parameters($searchParams));
+
+        if ('versions' !== $searchType) {
+            $this->saveSearchToHistory($results);
+        }
+
+        return $results;
+    }
+
+    /**
+     * Render page elements
+     *
+     * @param Params  $params  Request params
+     * @param Results $results Search results
+     *
+     * @return array
+     */
+    protected function getElements(Params $params, Results $results): array
+    {
+        if ($params->fromQuery('searchType', '') === 'versions') {
+            // Customize the URL helper to make sure it builds proper versions URLs
+            $results->getUrlQuery()
+                ->setDefaultParameter('id', $params->fromQuery('id'))
+                ->setDefaultParameter('keys', $params->fromQuery('keys'))
+                ->setSuppressQuery(true);
+        }
+
+        $result = [];
+        foreach ($this->elements as $selector => $element) {
+            $content = call_user_func([$this, $element['method']], $params, $results);
+            if (null !== $content) {
+                $result[$selector] = [
+                    'content' => $content,
+                    'target' => $element['target'],
+                    'attrs' => $element['attrs'] ?? [],
+                ];
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Render search results
+     *
+     * @param Params  $params  Request params
+     * @param Results $results Search results
+     *
+     * @return ?string
+     */
+    protected function renderResults(Params $params, Results $results): ?string
+    {
+        [$baseAction] = explode('-', $results->getOptions()->getSearchAction());
+        $templatePath = "$baseAction/results-list.phtml";
+        if ('search' !== $baseAction && !$this->renderer->resolver($templatePath)) {
+            $templatePath = 'search/results-list.phtml';
+        }
+        $params = $results->getParams();
+        $options = $results->getOptions();
+        $cart = $this->renderer->plugin('cart');
+        $showBulkOptions = $options->supportsCart()
+            && ($this->config['Site']['showBulkOptions'] ?? false);
+        // Checkboxes if appropriate:
+        $showCartControls = $options->supportsCart()
+            && $cart()->isActive()
+            && ($showBulkOptions || !$cart()->isActiveInSearch());
+        // Enable bulk options if appropriate:
+        $showCheckboxes = $showCartControls || $showBulkOptions;
```

Comment:
This feels like complex and potentially confusing logic that could cause problems if it gets out of sync. Is there any chance we could encapsulate it in a single place for sharing, or is that just not practical?

Reply:
Yes, and we can eliminate carrying these around everywhere. Now in #3292.

Reply:
Guess we should work on getting that finished up before continuing here!